Azure iot hub rest support

AieatAssam opened the pull request:
## Description:
Added a new component (Controller type) that automates posting device-to-cloud telemetry to Azure IoT Hub using REST API.
Requires Azure IoT Hub device provisioned (free tier hub is fine) to test.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#463

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

AieatAssam committed:
Added initial C++ code for supporting REST-based posting of events to Azure IoT Hub.

AieatAssam committed:
Added first pass of config validation and code generation

AieatAssam committed:
Fix to dump_config handling of strings

AieatAssam committed:
Removed ability to use IoT hub without security due to complexities involved in supporting insecure ESP32.
Tweaked configuration to better streamline differences between ESP32 and ESP8266.
Added test for iot hub.

AieatAssam committed:
Added debug logging of HTTP payload.
Resolved configuration and implementation issues.
Tested on real 8266 device successfully.

AieatAssam committed:
Heavily optimised code.
Baltimore root CA certificate is now stored in flash memory and loaded through #define.
JSON handling in hub packets is now using json_util.h to minimise impact on memory for string reallocations.
Hard-coded strings are changed to PROGMEM strings where applicable to reduce RAM impact.

AieatAssam committed:
Fixed typo in json property name.

AieatAssam committed:
Post linting cleanup to conform to coding standards

AieatAssam committed:
Improved parameters for test file.
Fixed pylint complaints about lines over 100 characters.

AieatAssam committed:
Restructured tests to deal with program space issues.

AieatAssam committed:
Added exception handling to better deal with SSL provisioning.

AieatAssam committed:
C++ cleanup to adhere to clang naming conventions enforced.
Pylint/flake8 bare except bypass for SSL certificate retrieval.

OttoWinter commented on esphome/components/azure_iot_hub/__init__.py:
```diff
@@ -0,0 +1,156 @@
+
+def retrieve_baltimore_root_ca():
+    cert_url = 'https://dl.cacerts.digicert.com/BaltimoreCyberTrustRoot.crt.pem'
+    with urlopen(cert_url) as response:
```
This will be run every time the user compiles, right?

I think in that case we should cache this result. We have a folder already for esphome internal data: `.esphome` (in the config dir, see also StorageJSON class).

Download it once, and cache it there. I guess there can also be problems if the root CA is updated (don't know how often that happens), but we probably can push that to later™.

AieatAssam replied:
Baltimore root CA changes very infrequently so could be cached. Ideally I'd like to introduce python [pem library](https://pem.readthedocs.io/en/stable/) to decode certificate and pull its expiration date out. This will permit better handling of token validity.

SHA1 fingerprint for ESP8266 (shame that we can't do this with ESP32) can be cached too, it has a set expiration date. However, if Azure renew their certificate prior to expiration date, there is no mechanism to regenerate it in cache that I know of. Perhaps allow an option of bypass cache as optional configuration setting?  

Root CA can be used on ESP8266, but in my experience, ESP8266 struggles with SSL certificate chain validation if it has a reasonably complex firmware loaded. SHA1 is a lot lighter on runtime memory.

AieatAssam replied:
Removed certificate and sha1 fingerprint hadnling for now. Will reintroduce once http_request is extended to support it.

OttoWinter commented on esphome/components/azure_iot_hub/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    cv.GenerateID(): cv.declare_id(AzureIoTHub),
+    cv.Required(CONF_HUB_NAME): cv.string,
+    cv.Required(CONF_DEVICE_ID): cv.string,
+    cv.Required(CONF_DEVICE_KEY): cv.string,
```
What does such a device key look like? Does it only have digits in it?

(Asking because if it begins with a `0`, then YAML will remove the leading zero and treat it as an integer)

AieatAssam replied:
Device key is a base 64 string, so could potentially start with a zero. Should documentation specify that this needs to be wrapped in quotes if it starts with a zero?

OttoWinter commented on esphome/components/azure_iot_hub/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    cg.add(var.set_iot_hub_device_id(config[CONF_DEVICE_ID]))
+
+    # set default expiration to approximately 100 years
+    token_expiration = time() + 100 * 365 * 24 * 60 * 60
```
I think it would make more sense to use the great `datetime` built-in library here (`datetime.utcnow() + timedelta(...)`)

AieatAssam replied:
This is a good point. I will adjust accordingly.

OttoWinter commented on esphome/components/azure_iot_hub/azure_iot_hub.cpp:
```diff
@@ -0,0 +1,353 @@
+    ESP_LOGD(TAG, "Using insecure SSL");
+  }
+#endif
+  this->http_client_.setReuse(true);
```
So ESPHome has an integrated HTTP system now, and you should use that here too (I don't want to have multiple HTTP client implementations around).

https://github.com/esphome/esphome/tree/dev/esphome/components/http_request

Anonym-tsk replied:
I added reuse only in telegram_bot component, but it still doesn't merged.
https://github.com/esphome/esphome/pull/919/files#diff-5f4f939d6141293f47821f7f174a00cdR17

AieatAssam replied:
I saw the http_request component and shamelessly "borrowed" some aspects of the http code from there. The reason why I didn't use it, however, is that it doesn't handle any SSL validation on ESP32. I was reluctant to leave SSL on ESP32 open to man-in-the-middle attack by default. 

Perhaps best course of action would be to have another pull request for http_request component to extend it to permit specifying root CA on ESP32 as an optional configuration?

Anonym-tsk replied:
You can add fingerprint and certificate validation to http_request component )

AieatAssam replied:
As a first pass, I am removing validation and using http_request component as is. Once that is in place and working, I can introduce fingerprint and CA validation in both places as one changeset with a separate pull request. Makes it easier to review and keeps changes in one place for now.

OttoWinter commented on esphome/components/azure_iot_hub/azure_iot_hub.cpp:
```diff
@@ -0,0 +1,353 @@
+    return;
+
+  std::string payload = json::build_json([this, sensor, state](JsonObject &root) {
+    root[F("deviceId")] = this->get_iot_hub_device_id();
```
```suggestion
    root["deviceId"] = this->get_iot_hub_device_id();
```

There's not much use in using `F()` for such short strings (the pointer itself is 4-byte long already).

Plus ESPHome doesn't have that much of an issue with program size as other firmwares by design.

AieatAssam replied:
I am used to trying to optimise every single byte of runtime memory if I can. Thus on ESP8266 and ESP32 libraries that support PROGMEM (which ArduinoJson does), I tend to store all constant strings on flash. If this is a concern due to additional processing and time taken to take them out into runtime, I will amend. As you said - the saving here is minimal.

OttoWinter commented on tests/test4.yaml:
```diff
@@ -0,0 +1,1501 @@
+esphome:
```
This file is huge and impossible to review. Please try to reduce its size.

Also, it looks like `test3` already checks this integration a bit already. More test files increase CI check times (unfortunately, we don't have a scalable solution for that yet).

AieatAssam replied:
The test is pretty much a copy of Test3 with azure iot added at the bottom to ensure it compiles with different sensor types enabled on ESP32 platform (couldn't extend other tests since this pushed program size over the limit). Wanted to make sure code behind #defines is tested (#define SENSOR, #define SWITCH etc).  I have also added ESP8266 architecture to test3.

I will add comments to yaml and drop it down to just a single sensor of each type to make it clearer what I am trying to achieve.

OttoWinter reviewed (commented):
Thank you for this contribution! From a code point of view this looks good! One major point though is that you should use the `http_request` base developd by @Anonym-tsk.

There's also another point here about whether we should allow these HTTP integrations. Previously, I was quite strongly against that (see also telegram integration PR by @Anonym-tsk), but after experimenting with a couple of things during the winter holidays I'm a bit more favorable towards HTTP integrations.

@brandond @glmnet @Anonym-tsk What do you guys think about allowing more HTTP integrations? Should we still draw a line for what is accepted (weather integrations might be one such example)?

Personally, I think a system where we add "code owners" to these integrations might be useful - that way, when triaging issues, you know who to tag and who should solve issues in that part of the system (similar to HA's issue tracker).

Anonym-tsk commented:
I have nothing against HTTP integrations. However, we should note users that such integrations require a lot of memory and can work unstably (especially on ESP8266).

"Code owners" system is good too.

AieatAssam commented:
Thank you for reviewing. I really like ESPHome codebase now that I got a chance to dig through it. Code generationaspects in particular are absolutely great. 
I started using esphome recently and found it to work great for me and take a lot of the work out of dealing with some standard gotchas of espressif devices (such as wifi chip on ESP8266 locking up and needing reboot after running for a while). 

Azure IoT Hub as IoT platfrorm is something I am working a lot with and I could see extending to support it being a fairly straightforward addition to the existing codebase. At the same time, MQTT support already in the platform can _theoretically_ be used for Azure IoT Hub communications rather than REST that I have implemented here. However, that requires secure MQTT which with the current codebase seems to only be supported on ESP8266 and even then fails to compile (there's outstanding PR and issues for this already).

Hence why I added the REST support. Admittedly, the memory requirements and stability of http(s) on ESP devices is an issue (as pointed out by @Anonym-tsk ), but it unlocks an additional platform for integrations and is something that is enabled on demand - i.e. if you don't need it, you don't have to use it. Amending documentation to make these issues clear is a good idea.

brandond commented:
I like the idea of putting some warnings on not just HTTP components, but any 3rd party controller integration. Obviously not all of us have every type of sensor, but we do at least use ESPHome with Home Assistant. If folks start adding additional controllers (Azure, AWS IoT, generic REST, etc) users should be aware that they might not work as well.

AieatAssam committed:
Changed Azure IoT Hub REST component to utilise existing http_request component.
Amended test4.yaml to make it leaner and clear what is is intended for.
Removed SSL security validation from Azure IoT Hub component due to these being currently unsupported by http_request component. These will be reintroduced after http_request supports them.
Improved sensor state handling by introducing accuracy-based rounding.
Removed FlashStringHelper overheads for small strings.

AieatAssam committed:
Lint and formatting suggestions.
Now using arduino framework base versions of roundf.

AieatAssam commented:
I have made tha alterations a while ago. Any chance to review these again/reopen? It is now utilising http component only and no longer has direct HttpClient communications.
